pkg/commands/root.go

```diff
@@ -1,160 +1,168 @@
 		RunE: func(cmd *cobra.Command, _ []string) error {
+			if c.opts.PrintVersion {
+				_ = printVersion(logutils.StdOut, info)
+				return nil
+			}
+
 			return cmd.Help()
 		},
-		PersistentPreRunE:  e.persistentPreRun,
-		PersistentPostRunE: e.persistentPostRun,
 	}
 
-	initRootFlagSet(rootCmd.PersistentFlags(), e.cfg)
-
-	e.rootCmd = rootCmd
-}
+	fs := rootCmd.Flags()
+	fs.BoolVar(&c.opts.PrintVersion, "version", false, color.GreenString("Print version"))
 
-func (e *Executor) persistentPreRun(_ *cobra.Command, _ []string) error {
-	if e.cfg.Run.PrintVersion {
-		_ = printVersion(logutils.StdOut, e.buildInfo)
-		os.Exit(exitcodes.Success) // a return nil is not enough to stop the process because we are inside the `preRun`.
-	}
+	setupRootPersistentFlags(rootCmd.PersistentFlags(), &c.opts)
 
-	runtime.GOMAXPROCS(e.cfg.Run.Concurrency)
+	reportData := &report.Data{}
+	log := report.NewLogWrapper(logutils.NewStderrLog(logutils.DebugKeyEmpty), reportData)
 
-	if e.cfg.Run.CPUProfilePath != "" {
-		f, err := os.Create(e.cfg.Run.CPUProfilePath)
-		if err != nil {
-			return fmt.Errorf("can't create file %s: %w", e.cfg.Run.CPUProfilePath, err)
-		}
-		if err := pprof.StartCPUProfile(f); err != nil {
-			return fmt.Errorf("can't start CPU profiling: %w", err)
-		}
-	}
+	// Dedicated configuration for each command to avoid side effects of bindings.
+	rootCmd.AddCommand(
+		newLintersCommand(log, config.NewDefault()).cmd,
+		newRunCommand(log, config.NewDefault(), reportData, info).cmd,
+		newCacheCommand().cmd,
+		newConfigCommand(log).cmd,
+		newVersionCommand(info).cmd,
+	)
 
-	if e.cfg.Run.MemProfilePath != "" {
-		if rate := os.Getenv(envMemProfileRate); rate != "" {
-			runtime.MemProfileRate, _ = strconv.Atoi(rate)
-		}
-	}
+	rootCmd.SetHelpCommand(newHelpCommand(log).cmd)
 
-	if e.cfg.Run.TracePath != "" {
-		f, err := os.Create(e.cfg.Run.TracePath)
-		if err != nil {
-			return fmt.Errorf("can't create file %s: %w", e.cfg.Run.TracePath, err)
-		}
-		if err = trace.Start(f); err != nil {
-			return fmt.Errorf("can't start tracing: %w", err)
-		}
-	}
+	c.log = log
+	c.cmd = rootCmd
 
-	return nil
+	return c
 }
 
-func (e *Executor) persistentPostRun(_ *cobra.Command, _ []string) error {
-	if e.cfg.Run.CPUProfilePath != "" {
-		pprof.StopCPUProfile()
+func (c *rootCommand) Execute() error {
+	err := setupLogger(c.log)
+	if err != nil {
+		return err
 	}
 
-	if e.cfg.Run.MemProfilePath != "" {
-		f, err := os.Create(e.cfg.Run.MemProfilePath)
-		if err != nil {
-			return fmt.Errorf("can't create file %s: %w", e.cfg.Run.MemProfilePath, err)
-		}
+	return c.cmd.Execute()
+}
 
-		var ms runtime.MemStats
-		runtime.ReadMemStats(&ms)
-		printMemStats(&ms, e.log)
+func setupRootPersistentFlags(fs *pflag.FlagSet, opts *rootOptions) {
+	fs.BoolVarP(&opts.Verbose, "verbose", "v", false, color.GreenString("Verbose output"))
+	fs.StringVar(&opts.Color, "color", "auto", color.GreenString("Use color when printing; can be 'always', 'auto', or 'never'"))
```

Comment:
are `verbose` and `color` actual for not `run` commands?

Reply:
they are global, but they are always used by the root command like before.